Code/GraphMol/Substruct/test1.cpp

```diff
@@ -1589,8 +1589,212 @@ void testGithubIssue1489() {
+      const auto query = R"([C@@])"_smarts;
+      std::vector<MatchVectType> matches;
+      TEST_ASSERT(!SubstructMatch(*mol, *query, matches, uniquify,
+                                  recursionPossible, useChirality));
+    }
+  }
+  {  // Mol also starting with the chiral atom
+    const auto mol = R"([C@](C)(Cl)(Br)F)"_smiles;
+    {
+      const auto query = R"([C@](C)(Cl)(Br)F)"_smarts;
+      std::vector<MatchVectType> matches;
+      TEST_ASSERT(SubstructMatch(*mol, *query, matches, uniquify,
+                                 recursionPossible, useChirality));
+    }
+    {
+      const auto query = R"([C@@](C)(Cl)(Br)F)"_smarts;
+      std::vector<MatchVectType> matches;
+      TEST_ASSERT(!SubstructMatch(*mol, *query, matches, uniquify,
+                                  recursionPossible, useChirality));
+    }
+    {
+      const auto query = R"([C@](C)(Cl)Br)"_smarts;
+      std::vector<MatchVectType> matches;
+      TEST_ASSERT(SubstructMatch(*mol, *query, matches, uniquify,
+                                 recursionPossible, useChirality));
+    }
+    {
+      const auto query = R"([C@@](C)(Cl)Br)"_smarts;
+      std::vector<MatchVectType> matches;
+      TEST_ASSERT(!SubstructMatch(*mol, *query, matches, uniquify,
+                                  recursionPossible, useChirality));
+    }
+    {
+      const auto query = R"([C@](Cl)(Br)F)"_smarts;
+      std::vector<MatchVectType> matches;
+      TEST_ASSERT(!SubstructMatch(*mol, *query, matches, uniquify,
+                                  recursionPossible, useChirality));
+    }
+    {
+      const auto query = R"([C@@](Cl)(Br)F)"_smarts;
+      std::vector<MatchVectType> matches;
+      TEST_ASSERT(SubstructMatch(*mol, *query, matches, uniquify,
+                                 recursionPossible, useChirality));
+    }
+    {  // Swap order of a pair of atoms
+      const auto query = R"([C@](Br)(Cl)F)"_smarts;
+      std::vector<MatchVectType> matches;
+      TEST_ASSERT(SubstructMatch(*mol, *query, matches, uniquify,
+                                 recursionPossible, useChirality));
+    }
+    {
+      const auto query = R"([C@@](Br)(Cl)F)"_smarts;
+      std::vector<MatchVectType> matches;
+      TEST_ASSERT(!SubstructMatch(*mol, *query, matches, uniquify,
+                                  recursionPossible, useChirality));
+    }
+  }
+
+  {  // Start from a physical H atom
+    const auto mol = R"([H][C@](O)(F)Cl)"_smiles;
+    const auto smarts = MolToSmarts(*mol);
+    const auto query = SmartsToMol(smarts);
+    TEST_ASSERT(smarts == R"([#8]-[#6@@H](-[#9])-[#17])");
+    std::vector<MatchVectType> matches;
+    TEST_ASSERT(SubstructMatch(*mol, *query, matches, uniquify,
+                               recursionPossible, useChirality));
+  }
+  {
+    const auto mol = R"([H][C@](O)(F)Cl)"_smiles;
+    const auto query = R"([C@H](O)(F)Cl)"_smarts;
+    std::vector<MatchVectType> matches;
+    TEST_ASSERT(!SubstructMatch(*mol, *query, matches, uniquify,
```

Comment:
I believe that this is incorrect. 
Since the SMARTS `[C@](O)(F)Cl` matches, `[C@H](O)(F)Cl` definitely should do so as well.

Reply:
Right. I don't know where my thoughts were on that one. I probably got confused with the matching between the neighboring H and the explicit H... somehow.

Reply:
I'm reviewing this, and thinking: how is this case different from the one in one of your previous comments:

> The SMILES `[C@H](F)(Cl)C` may need an inversion because the H atom is actually there in the molecule and "participating in defining the chirality".
> The SMARTS `[C@H](F)(Cl)C`, on the other hand, doesn't actually have an H affecting the chirality; it has a query that there's an H attached to the C.

This is almost exactly the same case, so according to that, this test was correct, and neither SMARTS [C@](O)(F)Cl nor [C@H](O)(F)Cl should match SMILES [C@H](O)(F)Cl, as the SMILES will have its chirality reversed??